### PR TITLE
Reveal block boundaries on hover in the Site Editor

### LINKED ISSUE #27239: Full Site Editing: Block selection
Anyone who has spent some time playing with the Site Editor – and Template Parts in particular – will likely empathise with the struggles one can face when interacting with these container blocks and their children. 

I frequently feel a distinct lack of confidence in any attempt I make to select a template part via click. More often than not I inadvertently select a child block, and from there I have to decide whether to:

* Persist with my clicking attempts (getting more frustrated with each failed click), 
* Utilise the block list view (ugh two clicks + visually scanning through a list), 
* Or admit defeat by selecting a child block and then using the parent selector pattern – multiple times if there's a lot of nesting. 

It probably shouldn't be this difficult to simply select a container block like a template part of a group.

---

Since the template editing experience in the Site Editor will largely revolve around interactions with container blocks like Template Parts, Queries, and Groups, I'm wondering if that environment might be a good space to explore solutions to this problem.

Doing so should also address the final high priority item in part 2 of the [Site Editing Infrastructure and UI](https://github.com/WordPress/gutenberg/issues/24818) milestone.

### A prototype

To tackle this head-on I put a rough prototype together which addresses the aforementioned issues via a combination of UI adjustments.

select

[Take it for a spin yourself](https://cdpn.io/jameskoster/debug/ZEOgLxV).

* Hovering any block will outline its boundary, helping clarify what will actually be selected when you click.
* Selecting a block removes the border, with the exception of container blocks as it is important to understand the outer boundaries of a container. This avoids borders distracting you from interacting with actual content.
* Template parts have a different border color to indicate that they are special. More context on this in https://github.com/WordPress/gutenberg/issues/26599
* When a child block is selected a permanently-visible shortcut to select its parent appears left of its toolbar. Hovering this shortcut reveal the parent boundary.
* If the selected child block is inside a Template Part, that Template Parts name is highlighted in the Top Bar, and can be clicked to navigate directly to the template part. This is helpful if the nesting is deep, or the layout is convoluted.

The prototype includes regular blocks, two separate template parts, and a group block as examples.

I appreciate we've had borders outline blocks similar to this in the past, and moved away from that pattern. But in the context of editing templates with lots of nested elements I find the indications to be very helpful and intuitive. I'm sure there is a lot of refinement still to do, particular around the visuals (border widths, color treatments, etc), but this feels like a decent starting point to me.

cc @jasmussen @shaunandrews @kjellr

### PR DESCRIPTION
**Pull request by jameskoster**
This PR roughly implements parts of a [prototype](https://cdpn.io/jameskoster/debug/ZEOgLxV) I worked on for https://github.com/WordPress/gutenberg/issues/27239, to see how it feels more holistically.

boundaries

Block boundaries are exposed on hover making it easier to understand what will be selected upon click. This is important in the Site Editor particularly, because block nesting features prominently in that context.

A different color is applied to template parts, to indicate that they are different to regular blocks.

This feels like a net positive change to me, although I would say that the `is-hovered` class seems to be somewhat brittle. Moving the cursor around quickly can trip it up. It doesn't occur all that often, but if there's any way to improve performance there it would be worth exploring.

To test:

* Load up a template in the Site Editor
* Apply hover, select, focus, select+focus states to various blocks and ensure the following:

1. Un-selected, hovered blocks display a 1px blue outline
2. Selected + focussed blocks (IE on-click) display a 1.5px blue outline
3. Selected un-focussed blocks display a 1px black outline

### CODE REVIEW

File: packages/block-editor/src/components/block-list/style.scss [hunk 1]
```diff
@@ -208,7 +208,7 @@
 
 	// Active entity spotlight.
 	&.has-active-entity:not(.is-focus-mode) {
-		opacity: 0.5;
+		//opacity: 0.5;
```
Comment (jasmussen) [inline packages/block-editor/src/components/block-list/style.scss:211]: Probably need to remove this when this PR goes out of draft state.

Comment (jameskoster) [reply packages/block-editor/src/components/block-list/style.scss:211]: Probably 😅

**Comment by jasmussen** [comment]
Thank you for this PR. As will be evident from the GIF below, my site editor is currently a little messy, so it's not totally accurate. But hopefully a few aspects can serve as legitimate feedback:

test

First off, the rendering issues with the 1.5px border are present here. I've a bit of tasks on my plate today, but I'm happy to try and dive in tomorrow and see if I can help fix it. 

Secondly, and as noted in https://github.com/WordPress/gutenberg/issues/27239, I think we should avoid different colors for template part boundaries. Right now we have black and blue colors (in the default color scheme), and those are sufficiently challenged on black and blue backgrounds, that I think we need a smarter system for rendering colors on arbitrary backgrounds before we introduce additional ones.

border

In this one, there's both some weirdness with the hover style (there's a dead zone). But also the fact that it looks like the border of the placeholder box grows thicker. This is probably a coincidence, because the hover is 1px outset from the border of the placeholder. 

I wonder if we should have a spot colored 1px hover style, instead of black? This would make blue the color for all things selected/focused:

- Hovered: 1px blue
- Selected and focused: 1.5px blue border
- Not part of this PR, but could be explored in the future: selected but not focused: 1px blue border.

What do you think?

**Comment by jameskoster** [comment]
Updated to focus on hover interactions only for now (removed template part colors).

I _think_ the selected / focused / hovered outlines are behaving as expected, but I am not 100% familiar with the interaction there so more eyes would be helpful. One question I have there is: which treatment takes priority? IE if I hover a selected item, which outline style should prevail, 1px (hovered) or 1.5px outline (selected)?

**Comment by jasmussen** [comment]
> One question I have there is: which treatment takes priority? IE if I hover a selected item, which outline style should prevail, 1px (hovered) or 1.5px outline (selected)?

There are no rules to follow here, so it's what we choose as best. Would it be useful for a focused block to have a hover effect at all?

Note, I figured out why the focus outline is weirdly cropped like so:

Screenshot 2020-11-26 at 10 41 18

If you add `outline: 2px solid transparent;` to that focus style, it looks right:

Screenshot 2020-11-26 at 10 40 31

The benefit is that this focus style will also be visible in windows 10 high contrast mode.

**Comment by jameskoster** [comment]
Updated so that the selected style persists over the hover style. Since the hover style intends to outline what will be selected on click, it's probably not adding much value when the block is already selected.

**Comment by jasmussen** [comment]
Current status:

status

This is looking like a good small step that I think is worth trying. And most of the issues appear to have been fixed.

It looks to me like some of the blue borders are used incorrectly in the site editor — that's the case also in the master branch so not the fault of this PR — but they linger even if focus moves elsewhere. The thick blue border is focus only, so we should look at refactoring it to actually be attached to `:focus`. We _can_ absolutely explore also having a border on the `.is-selected` version, but this border should be separate from focus — my instinct is 1px black. Jay if you get an opportunity to look at this, I'd appreciate it — otherwise I'll take a look when I have a chance.

**Comment by jameskoster** [comment]
I think I have it:

select

I'm also trying `cursor: default` on hover as I find it a little distracting that the cursor _and_ the outline change.

**Comment by mtias** [comment]
Thanks for exploring! We might want to leave the outline visible on select as well until is-typing kicks in.

**Comment by karmatosed** [comment]
In exploring this it feels good as I interact. The hitches come as mentioned where certain behaviours outside the scope of this clash against the hover. This seems outside the scope here, but I mention as it could do with easing to ensure the experience flows as it can distract.  For example:

image

Similarly here:

**Comment by jameskoster** [comment]
> Thanks for exploring! We might want to leave the outline visible on select as well until is-typing kicks in.

Done.

I've also tweaked the outline alignments further... getting there!

**Comment by jasmussen** [comment]
Seems related: #27075

**Comment by shaunandrews** [comment]
This is a solid first step.

I think the subtle differences between 1px, 1.5px, 3px, blue, and black borders is going to be misunderstood by most people.

I find the current 1px, solid, blue line distracting but am willing to live with it as we improve as it does solve a number of real problems. I'd suggest moving away from box-shadow hacks and rounded corners, and sticking with the simple CSS `outline` for all block borders. Then, I can see using a dotted outlined to represent the hover state in a more subtle way.

**Review by shaunandrews — APPROVED** [review]

File: packages/block-editor/src/components/block-list/style.scss [hunk 2]
```diff
@@ -208,7 +208,6 @@
 
 	// Active entity spotlight.
 	&.has-active-entity:not(.is-focus-mode) {
-		opacity: 0.5;
```
Comment (noahtallen) [inline packages/block-editor/src/components/block-list/style.scss:211]: If we are removing the template part spotlight mode thingy, we can probably remove this class entry altogether

**Review by noahtallen — COMMENTED** [review]
I like it!

For some reason, the template part border is the same color for me. If that's intentional, then I'm curious why we need extra classes to handle borders for the template part block specifically? It seems like it could just be generic. I also agree with shaun re: not using box-shadows, but I'm ok moving forward with this to see how folks like it!

**Comment by jameskoster** [comment]
> the template part border is the same color for me. If that's intentional, then I'm curious why we need extra classes to handle borders for the template part block specifically?

It's intentional for now, we might re-explore later. The only template part specific class I can see is to handle the offset on the :after, but it doesn't look like we really need that, so I'll remove for now.

**Comment by jameskoster** [comment]
Updated based on your review @noahtallen, thanks for that. I also moved the outline styles to what I believe is a more appropriate file.

**Comment by Addison-Stavlo** [comment]
Im curious why we are limiting this change to the site editor?  Is the idea to try it out here and potentially move these over to the post editors if it proves out well?  

It seems like a lot of the interactions this aims to improve are not site editor specific.  While it is noted above that block nesting is featured more prominently in this context by default, it can still be just as (or more) prominent in the post editors based on the users needs and designs that are being implemented there.

**Comment by noahtallen** [comment]
It'd be cool to have it as a setting. Maybe it would be off by default there, and on by default in the site editor. (maybe that could be a followup?)

**Comment by mtias** [comment]
By default, this effect could be jarring on the post editor alone (we tried it a while ago) overemphasizing blocks too much. I think eventually, if this iteration works out for the better, it should be part of the "select" tool. The site editor could make the select tool the default, then.

**Comment by jeyip** [comment]
### Testing
#### Requirements
I tested template parts and a variety of different blocks (image, paragraph, etc.)
- [x] Un-selected, hovered blocks display a 1px blue outline
- [x] Selected + focused blocks display a 1.5px blue outline (I clicked on a block directly)
- [x] Selected un-focussed blocks display a 1px black outline (I clicked a block and then clicked on block settings)

#### Browsers
- [x] Firefox
- [x] Chrome
- [x] IE11
- [x] Edge
- [x] Safari

I did notice a bit of wonky behavior with placeholders. @jasmussen [pointed out issues with dead zones earlier](https://github.com/WordPress/gutenberg/pull/27271#issuecomment-733714011), and I'm seeing them as well. 


Overall, I think the code changes look reasonable, and the editing experience, to me, definitely feels more intuitive than before. Thanks for working on this @jameskoster 🙌

**Comment by jameskoster** [comment]
Thanks for testing @jeyip :) The 'dead zone' issue is reported [here](https://github.com/WordPress/gutenberg/issues/27296).

**Pull request merged**